### PR TITLE
Parse type checking patterns in Java

### PR DESCRIPTION
**Pull request by emjin**
Test plan: test against these patterns
$T $B;
...
(String $B);
...
(String $B).func();
...
func ((typ $A));
...
$A = (typ $B);

### CODE REVIEW

**Comment by emjin** [comment]
Note that $A.field does not parse, with or without types

File: lang_java/parsing/parser_java.mly [hunk 1]
```diff
@@ -416,6 +416,12 @@ bound: ref_type_and_list { $1 }
 /*(*1 Expressions *)*/
 /*(*************************************************************************)*/
 
+typed_metavar:
+ | LP type_ IDENTIFIER RP { (match $2 with
```
Comment (nbrahms) [inline lang_java/parsing/parser_java.mly:420]: FWIW I'd probably put the first statement in an auxiliary function,

e.g.
```ocaml
let fail_if_metavar_on_type: type_ -> () =
  | TClass (((typ, _), _)::_) when typ.[0] = '$' -> failwith ...
  | _ -> ()
```

Then you can just call
`fail_if_metavar_on_type $2;`
here.

**Comment by nbrahms** [comment]
FWIW, you can drop pattern files in `tests/java/semgrep` to validate pattern parsing.

**Comment by emjin** [comment]
Limitations:
If A is a String[], (String $B) will not match A[i].
Field access doesn't always work. See https://github.com/returntocorp/semgrep/pull/930/files, typed_fieldaccess.

File: lang_java/parsing/parser_java.mly [hunk 2]
```diff
@@ -416,6 +423,10 @@ bound: ref_type_and_list { $1 }
 /*(*1 Expressions *)*/
 /*(*************************************************************************)*/
 
+typed_metavar:
+ | LP type_ IDENTIFIER RP { fail_if_metavar_on_typecheck $2;
+                            Flag_parsing.sgrep_guard (TypedMetavar($3, Parse_info.fake_info " ", $2))  }
```
Comment (aryx) [inline lang_java/parsing/parser_java.mly:428]: no need the fake_info here if you remove tok from TypedMetavar

File: lang_java/parsing/parser_java.mly [hunk 3]
```diff
@@ -826,6 +839,7 @@ statement_expression:
  | class_instance_creation_expression  { $1 }
  /*(* sgrep-ext: to allow '$S;' in sgrep *)*/
  | IDENTIFIER { Flag_parsing.sgrep_guard ((Name (name [Id $1])))  }
+ | typed_metavar { $1 }
```
Comment (aryx) [inline lang_java/parsing/parser_java.mly:842]: not sure we need this one.

File: lang_java/analyze/java_to_generic.ml [hunk 4]
```diff
@@ -300,6 +300,10 @@ and expr e =
   | AssignOp ((v1, (v2, tok), v3)) ->
       let v1 = expr v1 and v3 = expr v3 in
       G.AssignOp (v1, (v2, tok), v3)
+  | TypedMetavar (v1, v2, v3) ->
+      let v1 = ident v1 in
+      let v3 = typ v3 in
+      G.TypedMetavar (v1, v2, v3)
```
Comment (aryx) [inline lang_java/analyze/java_to_generic.ml:306]: here you can put a fake_parse_info for v2
and later one I'll remove the tok also in AST_generic.TypedMetavar.

File: lang_java/parsing/parser_java.mly (hunk 3 above)
Comment (emjin) [reply lang_java/parsing/parser_java.mly:842]: We need it so that we can match (String $A)

**Pull request merged**